### ISSUE TITLE
Additional 1s latency in `host -> service IP -> pod` with iptables

### ISSUE THREAD

**Issue by couloum** [kubernetes/kubernetes#96868]
**What happened**:

With a freshly installed Kubernetes, I observe exactly the same issue as the one described in the very well documented issue #90854

I applied a suggested fix of deactivating vxlan offloading and it worked correctly.

However, I use iptables in my setup and not ipvs. The issue talk about a permanent fix supplied to kernel, but it seems to be related to IPVS. My question is: is there a fix for iptables other than deactivating vxlan offloading?

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"8d8aa39598534325ad77120c120a22b3a990b5ea", GitTreeState:"clean", BuildDate:"2020-03-12T21:03:42Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.13", GitCommit:"30d651da517185653e34e7ab99a792be6a3d9495", GitTreeState:"clean", BuildDate:"2020-10-15T00:59:17Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: `proxmox-ve 6.2-1`
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
- Kernel (e.g. `uname -a`): 
```
Linux XXXX 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools: `kubeadm`
- Network plugin and version (if this is a network-related bug): 
```
# iptables --version
iptables v1.8.4 (legacy)

# calicoctl version
Client Version:    v3.17.0
Git commit:        0e7fc7f7
Cluster Version:   v3.16.4
Cluster Type:      k8s,kdd,bgp,kubeadm

# kubectl -n kube-system get cm calico-config -o yaml
...
data:
  calico_backend: vxlan
  cni_network_config: |-
    {
      "name": "k8s-pod-network",
      "cniVersion": "0.3.1",
      "plugins": [
        {
          "type": "calico",
          "log_level": "info",
          "log_file_path": "/var/log/calico/cni/cni.log",
          "datastore_type": "kubernetes",
          "nodename": "__KUBERNETES_NODE_NAME__",
          "mtu": __CNI_MTU__,
          "ipam": {
              "type": "calico-ipam"
          },
          "policy": {
              "type": "k8s"
          },
          "kubernetes": {
              "kubeconfig": "__KUBECONFIG_FILEPATH__"
          }
        },
        {
          "type": "portmap",
          "snat": true,
          "capabilities": {"portMappings": true}
        },
        {
          "type": "bandwidth",
          "capabilities": {"bandwidth": true}
        }
      ]
    }
  typha_service_name: none
  veth_mtu: "1450"
```
- Others:

**Comment by couloum** [comment]
I found my answer by reading more github issues here:
- https://github.com/projectcalico/calico/issues/3145
- https://github.com/coreos/flannel/pull/1282
- https://github.com/kubernetes/kubernetes/pull/92035

And finally, it seems there is a fix added in Kubernetes 1.8.5 (https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.18.md#bug-or-regression-4)

```
Fixes a problem with 63-second or 1-second connection delays with some VXLAN-based network plugins which was first widely noticed in 1.16 (though some users saw it earlier than that, possibly only with specific network plugins). If you were previously using ethtool to disable checksum offload on your primary network interface, you should now be able to stop doing that. (#92035, @danwinship) [SIG Network and Node]
```

I'll try that.

**Comment by couloum** [comment]
I tested with k8s v1.18.12 and calico v3.17, but the issue still persist (as balleon is reporting in https://github.com/projectcalico/calico/issues/3145)

I found an other workarround than disabling offloading. I described it here: https://github.com/projectcalico/calico/issues/3145#issuecomment-734234736

But there should be a way to disable random-fully in kube-proxy to fully solve the issue.

**Comment by couloum** [comment]
At this stage, I the only workarround I have is to disable offloading. However, I don't understand why the bug is still present with k8s v1.18.12, while this PR should have resolve it https://github.com/kubernetes/kubernetes/pull/92584

Does anyone have a clue here?

Thanks

**Comment by dcbw** [comment]
@couloum It looks like Ubuntu 20.04.1 that you referenced above still uses a 5.4 kernel. The patch that fixed the cases we root-caused landed in the 5.7 kernel:

https://github.com/torvalds/linux/commit/ea64d8d6c675c0bb712689b13810301de9d8f77a

That patch should no longer require offload to be turned off for interfaces handling the VXLAN traffic. Any chance you can try a later kernel somehow?

**Comment by couloum** [comment]
Thanks @dcbw . Unfortunately, I cannot use a custom kernel on these servers as they are for production and we stick with Ubuntu's official packages.
I was hopping that Ubuntu would have backport this patch in their kernel. I'll dig around on Ubuntu side.

**Comment by thockin** [comment]
What should we do with this  bug, now?

**Comment by couloum** [comment]
Hi,

I have confirmed the patch is integrated in latest Kernel version of Ubuntu 20.04. I will update the kernel and see if I no longer observe the issue.

**Comment by dcbw** [comment]
@couloum great; please test as soon as you can and let us know if the situation improves. Thanks!

**Comment by couloum** [comment]
Hi,

After more digging, I think I was on a wrong lead.

First, I updated to latest kernel on Ubuntu 20.04 and I observed the exact same issue.
Then, I looked at kernel source in Ubuntu.

The kernel patch was inserted in Ubuntu 20.04 kernel in May 2020 according to their git repository: https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/commit/net/netfilter/nf_nat_proto.c?id=96648eac2143d07bb2af0b4a8fdb1a87c9a72e71

We can see that it was already part of kernel 5.4.0-52.57 I was using when I opened the issue: https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/tree/net/netfilter/nf_nat_proto.c?h=Ubuntu-5.4.0-52.57#n65

So, I don't believe the patch actually work. At least, not in my case. I believe the bug is still here.

**Comment by dcbw** [comment]
@couloum so to summarize, the patch I linked is present in the kernel you are using, and does not seem to help the problem.

Can you confirm the following workarounds discussed above still work?

1) disable random-fully in kube-proxy solves the issue
2) disabling checksum offload solves the issue

What specific command are you using to disable offload, that makes it work?

**Comment by thockin** [comment]
ping @couloum

**Comment by couloum** [comment]
Sorry, I had a lot on my plate recently. I try this asap.

**Comment by mcmcghee** [comment]
@couloum You also have the option of using the 5.8 kernel by installing the [hardware enablement stack](https://wiki.ubuntu.com/Kernel/LTSEnablementStack) with the linux-generic-hwe-20.04 (or linux-virtual-hwe-20.04) package.

**Comment by danwinship** [comment]
/area kube-proxy

note that _either_ having a fixed kernel _or_ having a new-enough kubernetes would fix the known vxlan offload issue, so this sounds like it's something separate from the earlier bug.

**Comment by couloum** [comment]
@dcbw , sorry for the delay in my answer.

I confirm that disabling checksum offload DOES solves the issue. I use this command to disable checksum offload:
```
ethtool --offload vxlan.calico rx off tx off
```

Regarding your question about `random-fully`. I was able to manually test that, as described here: https://github.com/projectcalico/calico/issues/3145#issuecomment-734234736
Manually disabling `random-fully` solved the issue as well. However, I was not able to disable it permanently. Also, as I saw in changelog, `random-fully` was added for good reasons, so I preferred to disabled `checksum offloading` than `random-fully`.

Does that answer all of your questions?

Thanks
